keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,364 @@
+	NodePublishSecretRef *LocalObjectReference
+}
+```
+
+### Secret references
+Secret references declared in an inline CSI volume can only be used with namespaces from pods where they are referenced .  For inline usage, secret references are stored in `LocalObjectReference` values:
+* `LocalObjectReference` do not include a namespace reference.  This is to prevent reference to arbitrary namespace values.
+* The namespace reference will be extracted from the the pod spec at different phases of storage lifecycle by the Kubelet or external CSI compoennts.
+* The Kubelet and external CSI components must ensure secret references can only be used with the namespace from inline pod spec.
+
+
+### Specifying allowed inline drivers with `PodSecurityPolicy`
+To control which driver is allowed to be used within a pod spec, this design will update the `PodSecurityPolicy` to introduce `AllowedCSIDrivers` as shown below:
+
+```go
+  type PodSecurityPolicySpec struct {
+	// <snip>
+
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers used inline in a pod spec.  Empty or nil indicates that all
+	// CSI drivers may be used.  This parameter is effective only when the usage of the CSI plugin
+	// is allowed in the "Volumes" field.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
+  }
+
+  // AllowedCSIDriver represents a single CSI driver that is allowed to be used.
+  type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
+  }  
+```
+
+Value `PodSecurityPolicy.AllowedCSIDrivers` must be explicitly set with the names of CSI drivers that are allowed to be embedded within a pod spec.  An empty value means no CSI drivers are allowed to be specified inline inside a pod spec.
+
+### Inline CSI volume operation stages
+When a CSI driver is used in an inline context, it works slightly differently then when originated from PV/PVC.  As mentioned earlier, the inline volumes will participate in some, but not all volume operation stages with some limitations discussed here.
+
+#### Provision/deletion
+Volume provision and deletion will work differently for inline volumes. For persistent inline volumes, provision/deletion is handled as follows:
+* Persistent inline volumes do not participate in the provision and deletion phases.
+* Persistent inline volumes expect provision and deletion to be handled outside of the driver.
+
+For ephemeral CSI drivers:
+* Ephemeral drivers will not receive provision/deletion API calls (since these stages are driven by a PV/PVC/StorageClass).
+* Ephemeral CSI drivers will have to delay or combine any provisioning/deprovisioning operation during different phase.
+
+#### Attach/detach
+CSI uses API object `storage.VolumeAttachment` to track and manage attach/detach storage operation. Currently that object contains reference to an associated PV when the driver is not used inline.  However,  it must be extended to support information from `CSIVolumeSource` (see above) when an inline volume is being attached.
```

Comment:
is this section only relevant for persistent inline CSI volumes? see question below at the "Ephemeral inline drivers must ignore attachment requests" line

Reply:
yeah. Not needed for ephemeral I think.

Reply:
@liggitt that is correct. Only persistent inline drivers will participate in that phase.

Reply:
Clarified.